File: ports/java/docproc/src/main/kotlin/net/russiandocs/docproc/imaging/Contours.kt
package net.russiandocs.docproc.imaging

import org.opencv.core.Core
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.MatOfPoint
import org.opencv.core.MatOfPoint2f
import org.opencv.core.Point as CvPoint
import org.opencv.core.RotatedRect
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc
import kotlin.math.abs

/** A point in image coordinates. Double, because contours carry sub-pixel positions. */
public data class Pt(val x: Double, val y: Double)

/**
 * Contour extraction and the OpenCV calls around it.
 *
 * **Every OpenCV default argument the reference relies on is passed EXPLICITLY here.** That is CONVENTIONS
 * trap 15 and it is not pedantry: `cv2.convexHull(cnt)` defaults to `clockwise=False`, while the JVM
 * binding, gocv and OpenCvSharp all make the parameter required — so a port CHOOSES a value instead of
 * inheriting one. The Go port chose `true` and lost six pixels of canvas width on an internal-passport
 * spread, because the hull's ORIENTATION decides which vertices Douglas-Peucker keeps. Audit against the
 * Python call site, never against the binding's own defaults.
 */
public object Contours {

    private fun toMatOfPoint2f(points: List<Pt>): MatOfPoint2f =
        MatOfPoint2f(*points.map { CvPoint(it.x, it.y) }.toTypedArray())

    /** Otsu threshold. Returns the binary image and the chosen threshold value. */
    public fun thresholdOtsu(src: Image, invert: Boolean): Pair<Image, Double> {
        val gray = if (src.channels == 1) src.clone() else Io.toGray(src)
        gray.use {
            val binary = Mat()
            val type = (if (invert) Imgproc.THRESH_BINARY_INV else Imgproc.THRESH_BINARY) or
                Imgproc.THRESH_OTSU
            val value = Imgproc.threshold(gray.mat, binary, 0.0, 255.0, type)
            return Image.wrap(binary) to value
        }
    }

    /** External contours only, no hierarchy — `RETR_EXTERNAL` with `CHAIN_APPROX_SIMPLE`. */
    public fun findExternalContours(src: Image): List<List<Pt>> {
        val contours = ArrayList<MatOfPoint>()
        val hierarchy = Mat()
        try {
            Imgproc.findContours(src.mat, contours, hierarchy, Imgproc.RETR_EXTERNAL,
                Imgproc.CHAIN_APPROX_SIMPLE)
            return contours.map { contour ->
                contour.toArray().map { Pt(it.x, it.y) }
            }
        } finally {
            hierarchy.release()
            contours.forEach { it.release() }
        }
    }

    /** Contour area, via the shoelace formula OpenCV uses. */
    public fun contourArea(points: List<Pt>): Double {
        if (points.size < 3) {
            return 0.0
        }
        val mat = toMatOfPoint2f(points)
        try {
            return abs(Imgproc.contourArea(mat))
        } finally {
            mat.release()
        }
    }

    /**
     * The convex hull. `clockwise = false` — the reference's default, and NOT the binding's.
     *
     * The JVM binding returns INDICES rather than points, which is a second trap on top of the
     * orientation one: `Imgproc.convexHull` fills a `MatOfInt`, and the caller has to map them back. A
     * port that expected points would find the "hull" is a list of small integers and, worse, might use
     * them as coordinates.
     */
    public fun convexHull(points: List<Pt>): List<Pt> {
        if (points.isEmpty()) {
            return emptyList()
        }
        // Integer points, because the JVM's convexHull takes MatOfPoint (int) rather than MatOfPoint2f.
        // The reference feeds it a contour, which is integral already, so nothing is lost here.
        val input = MatOfPoint(*points.map { CvPoint(it.x, it.y) }.toTypedArray())
        val indices = org.opencv.core.MatOfInt()
        try {
            Imgproc.convexHull(input, indices, false)
            val source = input.toArray()
            return indices.toArray().map { Pt(source[it].x, source[it].y) }
        } finally {
            input.release()
            indices.release()
        }
    }

    /** Perimeter of a CLOSED contour — `cv2.arcLength(pts, True)`. */
    public fun arcLength(points: List<Pt>): Double {
        if (points.size < 2) {
            return 0.0
        }
        val mat = toMatOfPoint2f(points)
        try {
            return Imgproc.arcLength(mat, true)
        } finally {
            mat.release()
        }
    }

    /** Douglas-Peucker simplification of a CLOSED contour. */
    public fun approxPolyDp(points: List<Pt>, epsilon: Double): List<Pt> {
        val input = toMatOfPoint2f(points)
        val output = MatOfPoint2f()
        try {
            Imgproc.approxPolyDP(input, output, epsilon, true)
            return output.toArray().map { Pt(it.x, it.y) }
        } finally {
            input.release()
            output.release()
        }
    }

    /**
     * The four corners of the minimum-area rectangle.
     *
     * Taken from `boxPoints` rather than reconstructed from centre/size/angle: OpenCV changed the
     * `minAreaRect` angle convention around 4.5, and a hand-rolled version silently produces the corners
     * in a different ORDER — which then feeds a perspective transform.
     */
    public fun minAreaRectPoints(points: List<Pt>): List<Pt> {
        val mat = toMatOfPoint2f(points)
        val box = Mat()
        try {
            val rect: RotatedRect = Imgproc.minAreaRect(mat)
            Imgproc.boxPoints(rect, box)
            return (0 until box.rows()).map { r ->
                Pt(box.get(r, 0)[0], box.get(r, 1)[0])
            }
        } finally {
            mat.release()
            box.release()
        }
    }

    /** Warps a quadrilateral onto an axis-aligned rectangle of the given size. */
    public fun warpPerspectiveQuad(src: Image, quad: List<Pt>, width: Int, height: Int): Image {
        require(quad.size == 4) { "imaging: warp needs 4 points, got ${quad.size}" }

        val source = toMatOfPoint2f(quad)
        val destination = MatOfPoint2f(
            CvPoint(0.0, 0.0),
            CvPoint((width - 1).toDouble(), 0.0),
            CvPoint((width - 1).toDouble(), (height - 1).toDouble()),
            CvPoint(0.0, (height - 1).toDouble()),
        )
        val transform = Imgproc.getPerspectiveTransform(source, destination)
        try {
            val dst = Mat()
            Imgproc.warpPerspective(src.mat, dst, transform,
                Size(width.toDouble(), height.toDouble()))
            return Image.wrap(dst)
        } finally {
            source.release()
            destination.release()
            transform.release()
        }
    }

    /** Horizontal concatenation. Heights must match. */
    public fun hStack(a: Image, b: Image): Image {
        require(a.height == b.height) {
            "imaging: hstack heights differ: ${a.height} vs ${b.height}"
        }
        val dst = Mat()
        Core.hconcat(listOf(a.mat, b.mat), dst)
        return Image.wrap(dst)
    }

    /** Vertical concatenation. Widths must match. */
    public fun vStack(a: Image, b: Image): Image {
        require(a.width == b.width) {
            "imaging: vstack widths differ: ${a.width} vs ${b.width}"
        }
        val dst = Mat()
        Core.vconcat(listOf(a.mat, b.mat), dst)
        return Image.wrap(dst)
    }

    /**
     * Rotates around a FRACTIONAL centre.
     *
     * **This is where D-08 does not apply to this port.** `Imgproc.getRotationMatrix2D` takes a
     * double-valued `Point`, so `(sw / 2.0, sh / 2.0)` is expressible and the real OpenCV function can be
     * called. gocv takes an integer `image.Point` and cannot, which is why the Go port builds the matrix
     * by hand — and its naive integer version was measured to shift the deskewer's variance array by
     * 3.8e-3 relative, above the 1e-3 policy.
     */
    public fun rotateAround(src: Image, centreX: Double, centreY: Double, angleDeg: Double): Image {
        val transform = Imgproc.getRotationMatrix2D(CvPoint(centreX, centreY), angleDeg, 1.0)
        try {
            val dst = Mat()
            Imgproc.warpAffine(src.mat, dst, transform,
                Size(src.width.toDouble(), src.height.toDouble()))
            return Image.wrap(dst)
        } finally {
            transform.release()
        }
    }

}

File: ports/dotnet/src/RussianDocs.DocumentProcessing/Imaging/Contours.cs
using OpenCvSharp;

namespace RussianDocs.DocumentProcessing.Imaging;

/// <summary>A point in image coordinates. Double, because contours carry sub-pixel positions.</summary>
public readonly record struct Point(double X, double Y);

/// <summary>
/// Contour extraction and the OpenCV calls around it.
///
/// <para>
/// **Every OpenCV default argument the reference relies on is passed EXPLICITLY here.** That is
/// CONVENTIONS trap 15 and it is not pedantry: <c>cv2.convexHull(cnt)</c> defaults to
/// <c>clockwise=False</c>, while OpenCvSharp, gocv and the JVM bindings all make the parameter
/// required — so a port CHOOSES a value instead of inheriting one. The Go port chose <c>true</c> and
/// lost six pixels of canvas width on an internal-passport spread, because the hull's ORIENTATION
/// decides which vertices Douglas-Peucker keeps. Audit against the Python call site, never against
/// the binding's own defaults.
/// </para>
/// </summary>
public static class Contours
{
    public static Point[] ToPoints(IEnumerable<Point2f> points) =>
        [.. points.Select(p => new Point(p.X, p.Y))];

    private static Point2f[] ToPoint2f(IReadOnlyList<Point> points) =>
        [.. points.Select(p => new Point2f((float)p.X, (float)p.Y))];

    private static OpenCvSharp.Point[] ToCvPoints(IReadOnlyList<Point> points) =>
        [.. points.Select(p => new OpenCvSharp.Point((int)p.X, (int)p.Y))];

    /// <summary>Otsu threshold. Returns the binary image and the chosen threshold value.</summary>
    public static (Image Binary, double Threshold) ThresholdOtsu(Image src, bool invert)
    {
        using Image gray = src.Channels == 1 ? src.Clone() : Io.ToGray(src);
        var binary = new Mat();
        ThresholdTypes type = (invert ? ThresholdTypes.BinaryInv : ThresholdTypes.Binary)
            | ThresholdTypes.Otsu;
        double value = Cv2.Threshold(gray.Mat, binary, 0, 255, type);
        return (Image.Wrap(binary), value);
    }

    /// <summary>
    /// External contours only, no hierarchy — <c>RETR_EXTERNAL</c> with <c>CHAIN_APPROX_SIMPLE</c>.
    /// </summary>
    public static List<Point[]> FindExternalContours(Image src)
    {
        Cv2.FindContours(src.Mat, out OpenCvSharp.Point[][] contours, out _,
            RetrievalModes.External, ContourApproximationModes.ApproxSimple);
        return [.. contours.Select(c => c.Select(p => new Point(p.X, p.Y)).ToArray())];
    }

    /// <summary>Contour area, via the shoelace formula OpenCV uses.</summary>
    public static double ContourArea(IReadOnlyList<Point> points) =>
        points.Count < 3 ? 0 : Math.Abs(Cv2.ContourArea(ToPoint2f(points)));

    /// <summary>
    /// The convex hull. <c>clockwise: false</c> — the reference's default, and NOT the binding's.
    /// </summary>
    public static Point[] ConvexHull(IReadOnlyList<Point> points)
    {
        if (points.Count == 0)
        {
            return [];
        }
        // The Point2f overload always returns points, so only the orientation needs saying — and it
        // is the one that matters. OpenCvSharp has no default for it, unlike cv2.
        Point2f[] hull = Cv2.ConvexHull(ToPoint2f(points), clockwise: false);
        return ToPoints(hull);
    }

    /// <summary>Perimeter of a CLOSED contour — <c>cv2.arcLength(pts, True)</c>.</summary>
    public static double ArcLength(IReadOnlyList<Point> points) =>
        points.Count < 2 ? 0 : Cv2.ArcLength(ToPoint2f(points), closed: true);

    /// <summary>Douglas-Peucker simplification of a CLOSED contour.</summary>
    public static Point[] ApproxPolyDp(IReadOnlyList<Point> points, double epsilon) =>
        ToPoints(Cv2.ApproxPolyDP(ToPoint2f(points), epsilon, closed: true));

    /// <summary>
    /// The four corners of the minimum-area rectangle.
    ///
    /// <para>
    /// Taken from <c>BoxPoints</c> rather than reconstructed from centre/size/angle: OpenCV changed
    /// the <c>minAreaRect</c> angle convention around 4.5, and a hand-rolled version silently
    /// produces the corners in a different ORDER — which then feeds a perspective transform.
    /// </para>
    /// </summary>
    public static Point[] MinAreaRectPoints(IReadOnlyList<Point> points)
    {
        RotatedRect rect = Cv2.MinAreaRect(ToPoint2f(points));
        return ToPoints(Cv2.BoxPoints(rect));
    }

    /// <summary>Warps a quadrilateral onto an axis-aligned rectangle of the given size.</summary>
    public static Image WarpPerspectiveQuad(Image src, IReadOnlyList<Point> quad, int width, int height)
    {
        if (quad.Count != 4)
        {
            throw new ArgumentException($"imaging: warp needs 4 points, got {quad.Count}");
        }

        Point2f[] source = ToPoint2f(quad);
        Point2f[] destination =
        [
            new(0, 0),
            new(width - 1, 0),
            new(width - 1, height - 1),
            new(0, height - 1),
        ];

        using Mat transform = Cv2.GetPerspectiveTransform(source, destination);
        var dst = new Mat();
        Cv2.WarpPerspective(src.Mat, dst, transform, new Size(width, height));
        return Image.Wrap(dst);
    }

    /// <summary>Horizontal concatenation. Heights must match.</summary>
    public static Image HStack(Image a, Image b)
    {
        if (a.Height != b.Height)
        {
            throw new ArgumentException($"imaging: hstack heights differ: {a.Height} vs {b.Height}");
        }
        var dst = new Mat();
        Cv2.HConcat([a.Mat, b.Mat], dst);
        return Image.Wrap(dst);
    }

    /// <summary>Vertical concatenation. Widths must match.</summary>
    public static Image VStack(Image a, Image b)
    {
        if (a.Width != b.Width)
        {
            throw new ArgumentException($"imaging: vstack widths differ: {a.Width} vs {b.Width}");
        }
        var dst = new Mat();
        Cv2.VConcat([a.Mat, b.Mat], dst);
        return Image.Wrap(dst);
    }
}
